# Exercici 1
Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

In [187]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Titulos=['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
housing = pd.read_csv('C:\\RAUL\\NISSAN\\GENERAL\\USB\\FORMACION\\BootCamp - Data Scientist_Analist\\IT Academy Cibernarium - Data Science\\Data Science Itinerario\\5.- Introducción al Test de Hipótesis y al Machine Learning\\Entrega\\housing data.csv', header=None, names=Titulos)
housing.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [69]:
#Separem les columnes independets de la comlumna dependent o target (MEDV) i fem l'split de Train/Test
from sklearn.model_selection import train_test_split

X = housing.drop(['MEDV'],axis=1).values
y = housing['MEDV'].values
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=0.3, random_state=0)

#### Models de Regressió

In [70]:
# Multiple Linear Regression
from sklearn.linear_model import LinearRegression

model_LinRegre = LinearRegression().fit(X_train, y_train)
y_pred_test_linear = model_LinRegre.predict(X_test)

In [71]:
# Polynomial Regression
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

poly=PolynomialFeatures(degree=2)
X_poly_train = poly.fit_transform(X_train) # Primer fem la transformació polinòmica
X_poly_test = poly.fit_transform(X_test) # També tranformem les dades de test 

model_Poly = LinearRegression().fit(X_poly_train, y_train) # i després fem el model 
y_pred_test_Poly = model_Poly.predict(X_poly_test)

In [72]:
# Decission Tree
from sklearn.tree import DecisionTreeRegressor

model_DecTree = DecisionTreeRegressor(random_state=0).fit(X_train, y_train)
y_pred_test_decission = model_DecTree.predict(X_test)

In [135]:
# Random Forest
from sklearn.ensemble import RandomForestRegressor

model_RanForest = RandomForestRegressor(n_estimators = 1000, random_state = 42).fit(X_train, y_train)
y_pred_test_RanForest = model_RanForest.predict(X_test)

In [136]:
# Neural Network
from sklearn.neural_network import MLPRegressor

model_NeuNet = MLPRegressor(random_state=1, max_iter=1000).fit(X_train,y_train)
y_pred_test_NeuNet = model_NeuNet.predict(X_test)

# Exercici 2
Compara’ls en base al MSE i al R2.

In [137]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [320]:
print("Test Set Regression Models Performance:\n")

print("** MSE (Mean Squared Error) **:")
print("Multiple Lineal Regression: ",mean_squared_error(y_test, y_pred_test_linear).round(4))
print("Polynomial Regression: ",mean_squared_error(y_test, y_pred_test_Poly).round(4))
print("Desiccion Tree: ", mean_squared_error(y_test, y_pred_test_decission).round(4))
print("Random Forest: ", mean_squared_error(y_test, y_pred_test_RanForest).round(4))
print("Neural Network: ", mean_squared_error(y_test, y_pred_test_NeuNet).round(4))

print("\n** R2 (Coefficient of determination) **:")
print("Multiple Lineal Regression: ",model_LinRegre.score(X_test, y_test).round(4))
print("Polynomial Regression: ",model_Poly.score(X_poly_test, y_test).round(4))
print("Desiccion Tree: ",model_DecTree.score(X_test, y_test).round(4))
print("Random Forest: ",model_RanForest.score(X_test, y_test).round(4))
print("Neural Network: ",model_NeuNet.score(X_test, y_test).round(4))

Test Set Regression Models Performance:

** MSE (Mean Squared Error) **:
Multiple Lineal Regression:  27.196
Polynomial Regression:  29.0042
Desiccion Tree:  26.352
Random Forest:  14.327
Neural Network:  24.6819

** R2 (Coefficient of determination) **:
Multiple Lineal Regression:  0.6734
Polynomial Regression:  0.6517
Desiccion Tree:  0.6835
Random Forest:  0.8279
Neural Network:  0.7036


# Exercici 3
Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

In [308]:
# Multiple Linear Regression
import warnings
warnings.filterwarnings("ignore")

model_LinRegre_2 = LinearRegression(normalize=True).fit(X_train, y_train)
y_pred_test_linear_2 = model_LinRegre_2.predict(X_test)

In [309]:
# Polynomial Regression
poly_2=PolynomialFeatures(degree=(2), order='F', interaction_only=True, include_bias=False)
X_poly_train_2 = poly_2.fit_transform(X_train) # Primer fem la transformació polinòmica
X_poly_test_2 = poly_2.fit_transform(X_test) # També tranformem les dades de test 

model_Poly_2 = LinearRegression().fit(X_poly_train_2, y_train) # i després fem el model 
y_pred_test_Poly_2 = model_Poly_2.predict(X_poly_test_2)

In [310]:
# Decission Tree
model_DecTree_2 = DecisionTreeRegressor(random_state=0, min_samples_leaf=11).fit(X_train, y_train)
y_pred_test_decission_2 = model_DecTree_2.predict(X_test)

In [311]:
# Random Forest
model_RanForest_2 = RandomForestRegressor(n_estimators = 1040, random_state = 42, min_weight_fraction_leaf=0.00005).fit(X_train, y_train)
y_pred_test_RanForest_2 = model_RanForest_2.predict(X_test)

In [312]:
# Neural Network
model_NeuNet_2 = MLPRegressor(random_state=1, max_iter=3300, solver='lbfgs', alpha=0.05).fit(X_train,y_train)
y_pred_test_NeuNet_2 = model_NeuNet.predict(X_test)

In [321]:
print("Test set Regression Models Performance with tuned parmeters:\n")

print("** MSE (Mean Squared Error) **:")
print("Multiple Lineal Regression: ",mean_squared_error(y_test, y_pred_test_linear_2).round(4))
print("Polynomial Regression: ",mean_squared_error(y_test, y_pred_test_Poly_2).round(4))
print("Desiccion Tree: ", mean_squared_error(y_test, y_pred_test_decission_2).round(4))
print("Random Forest: ", mean_squared_error(y_test, y_pred_test_RanForest_2).round(4))
print("Neural Network: ", mean_squared_error(y_test, y_pred_test_NeuNet_2).round(4))

print("\n** R2 (Coefficient of determination) **:")
print("Multiple Lineal Regression: ",model_LinRegre_2.score(X_test, y_test).round(4))
print("Polynomial Regression: ",model_Poly_2.score(X_poly_test_2, y_test).round(4))
print("Desiccion Tree: ",model_DecTree_2.score(X_test, y_test).round(4))
print("Random Forest: ",model_RanForest_2.score(X_test, y_test).round(4))
print("Neural Network: ",model_NeuNet_2.score(X_test, y_test).round(4))

Test set Regression Models Performance with tuned parmeters:

** MSE (Mean Squared Error) **:
Multiple Lineal Regression:  27.196
Polynomial Regression:  21.3951
Desiccion Tree:  22.2076
Random Forest:  14.2963
Neural Network:  24.6819

** R2 (Coefficient of determination) **:
Multiple Lineal Regression:  0.6734
Polynomial Regression:  0.743
Desiccion Tree:  0.7333
Random Forest:  0.8283
Neural Network:  0.7649


**Conclusions:**
* Multiple Linear Regression --> apliquem normalització dels camps però no aconseguim cap millora
* Polynomial Regression --> modificant els paràmetres per defecte del PolynomialFeatures aconseguim millorar el R2 de 0.65   a 0.74
* Decission Tree --> el paràmetre que millor resultat dona és el "min_samples_leaf", limitant les mostres per fulla, passem d'un R2 de 0.68 a 0.73. 
* Random Forest --> Ajustant paràmetres acoseguim una millora molt petita de 0.8279 a 0.8283
* Neural Network --> Incrementant el nombre d'itercions, canviant el solver i pujant el valor d'alpha aconseguim millorar R2 de 0.70 a 0.76

# Exercici 4
Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).

In [301]:
# Aplicarem el Grid Search per cadascun dels 5 models de regressió que estem probant.
from sklearn.model_selection import GridSearchCV

In [302]:
# Multiple Linear Regression
parameters_LR = {
            'normalize': [True, False],
            'fit_intercept': [True, False],
            'positive': [True, False]
            }
LR=LinearRegression()

grid_search_LR = GridSearchCV(LR, parameters_LR, scoring='r2', cv=5)
grid_search_LR.fit(X_train, y_train)
Y_test_predict_GS_LR = grid_search_LR.predict(X_test)

In [303]:
# Polynomial Regression
poly=PolynomialFeatures(degree=2)
X_poly_train = poly.fit_transform(X_train) # Primer fem la transformació polinòmica
X_poly_test = poly.fit_transform(X_test) # També tranformem les dades de test 

parameters_PR = {
            'normalize': [True, False],
            'fit_intercept': [True, False],
            'positive': [True, False]
            }
PR=LinearRegression()

grid_search_PR = GridSearchCV(PR, parameters_PR, scoring='r2', cv=5)
grid_search_PR.fit(X_poly_train, y_train)
Y_test_predict_GS_PR = grid_search_PR.predict(X_poly_test)

In [304]:
# Decission Tree
parameters_DT = {
                "criterion": ['squared_error', 'friedman_mse', 'poisson'],
                "splitter": ["best", "random"],
                "max_depth": range(1,8,1),
                "min_samples_split": range(2,11,1),
                "min_samples_leaf": range(1,30,5),
                }
DT=DecisionTreeRegressor(random_state = 0) #Fem servir la mateixa semilla que els anteriors

grid_search_DT = GridSearchCV(DT, parameters_DT, scoring='r2', cv=5)
grid_search_DT.fit(X_train, y_train)
Y_test_predict_GS_DT = grid_search_DT.predict(X_test)

In [305]:
# Random Forest
parameters_RF = {
                "max_depth": [3, None],
                "max_features": range(1,11,1),
                "min_samples_split": range(2,11,1),
                "bootstrap": [True, False],
                }
RF=RandomForestRegressor(random_state = 42)

grid_search_RF = GridSearchCV(RF, parameters_RF, scoring='r2', cv=5)
grid_search_RF.fit(X_train, y_train)
Y_test_predict_GS_RF = grid_search_RF.predict(X_test)

In [306]:
# Neural Network
parameters_NN = {
                "hidden_layer_sizes": [(50,50,50), (50,100,50), (100,1)], 
                "activation": ["identity", "logistic", "tanh", "relu"], 
                "solver": ["lbfgs", "sgd", "adam"], 
                "alpha": [0.00005,0.0005, 0.0001],
                "learning_rate": ['constant', 'invscaling', 'adaptive'],
                "warm_start" : [True, False]
                }

NN=MLPRegressor(random_state = 0, max_iter=1000)

grid_search_NN = GridSearchCV(NN, parameters_NN, scoring='r2', cv=5, n_jobs=-1)
grid_search_NN.fit(X_train, y_train)
Y_test_predict_GS_NN = grid_search_NN.predict(X_test)

In [322]:
print("Test Set Regression Models Performance applying GridSearchCV:\n")

print("** MSE (Mean Squared Error) **:")
print("Multiple Linear Regression GS: ", mean_squared_error(y_test, Y_test_predict_GS_LR).round(4))
print("Polynomial Regression GS: ", mean_squared_error(y_test, Y_test_predict_GS_PR).round(4))
print("Decission Tree GS: ", mean_squared_error(y_test, Y_test_predict_GS_DT).round(4))
print("Random Forest GS: ", mean_squared_error(y_test, Y_test_predict_GS_RF).round(4))
print("Neural Network: ", mean_squared_error(y_test, Y_test_predict_GS_NN).round(4))

print("\n** R2 (Coefficient of determination) **:")
print("Multiple Linear Regression GS: ",grid_search_LR.score(X_test, y_test).round(4))
print("Polynomial Regression GS: ",grid_search_PR.score(X_poly_test, y_test).round(4))
print("Decission Tree GS: ",grid_search_DT.score(X_test, y_test).round(4))
print("Random Forest GS: ",grid_search_RF.score(X_test, y_test).round(4))
print("Neural Network: ",grid_search_NN.score(X_test, y_test).round(4))

Test Set Regression Models Performance applying GridSearchCV:

** MSE (Mean Squared Error) **:
Linear Regression GS:  27.196
Polynomial Regression GS:  29.2525
Decission Tree GS:  31.3472
Random Forest GS:  19.7556
Neural Network:  21.5402

** R2 (Coefficient of determination) **:
Linear Regression GS:  0.6734
Polynomial Regression GS:  0.6487
Decission Tree GS:  0.6235
Random Forest GS:  0.7627
Neural Network:  0.7413


**Conclusions:**

A banda del Multiple Linear Regression que es manté igual, els altres 4 models empitjoren respecte als models dels exercicis 2 i 3, posiblement degut a que aquest primers models ténen overfitting i aquest amb el Gridsearch

**How to get Best Model**

In [337]:
print("** GridSearchCV Best Scores **:\n")
print("Best Score Linear Regression: ",grid_search_LR.best_score_.round(4))
print("Best Score Polynomial Regression: ",grid_search_PR.best_score_.round(4))
print("Best Score Decission Tree: ",grid_search_DT.best_score_.round(4))
print("Best Score Random Forest: ",grid_search_RF.best_score_.round(4))
print("Best Score Neural Network: ",grid_search_NN.best_score_.round(4))

** GridSearchCV Best Scores **:

Best Score Linear Regression:  0.7183
Best Score Polynomial Regression:  0.7274
Best Score Decission Tree:  0.788
Best Score Random Forest:  0.8807
Best Score Neural Network:  0.7783


In [346]:
DT_best=DecisionTreeRegressor(**grid_search_DT.best_params_).fit(X_train, y_train)
DT_best.score(X_test, y_test).round(4)

0.6892

In [344]:
RF_best=RandomForestRegressor(**grid_search_RF.best_params_).fit(X_train, y_train)
RF_best.score(X_test, y_test).round(4)

0.7446

# Exercici 5
No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

In [248]:
#Treiem la columna RM del dataframe per poder repitir després la resta de pasos dels Exercicis 1 i 2.
housing_sRM=housing.drop(['RM'], axis=1)
X_sRM = housing_sRM.drop(['MEDV'],axis=1).values
y_sRM = housing_sRM['MEDV'].values
X_sRM_train, X_sRM_test, y_sRM_train, y_sRM_test = train_test_split(X_sRM, y_sRM, test_size=0.3, random_state=0)

In [249]:
# Multiple Linear Regression --> apliquem normalització dels camps
import warnings
warnings.filterwarnings("ignore")

model_LinRegre_sRM = LinearRegression().fit(X_sRM_train, y_sRM_train)
y_pred_test_linear_sRM = model_LinRegre_sRM.predict(X_sRM_test)

In [250]:
# Polynomial Regression --> 
poly_sRM=PolynomialFeatures(degree=2)
X_sRM_poly_train = poly_sRM.fit_transform(X_sRM_train) # Primer fem la transformació polinòmica
X_sRM_poly_test = poly_sRM.fit_transform(X_sRM_test) # També tranformem les dades de test 

model_sRM_Poly = LinearRegression().fit(X_sRM_poly_train, y_sRM_train) # i després fem el model 
y_pred_test_Poly_sRM = model_sRM_Poly.predict(X_sRM_poly_test)

In [251]:
# Decission Tree
model_DecTree_sRM = DecisionTreeRegressor(random_state=0).fit(X_sRM_train, y_sRM_train)
y_pred_test_decission_sRM = model_DecTree_sRM.predict(X_sRM_test)

In [252]:
# Random Forest
model_RanForest_sRM = RandomForestRegressor(n_estimators = 3000, random_state = 1).fit(X_sRM_train, y_sRM_train)
y_pred_test_RanForest_sRM = model_RanForest_sRM.predict(X_sRM_test)

In [253]:
# Neural Network
model_NeuNet_sRM = MLPRegressor(random_state=1, max_iter=2000).fit(X_sRM_train, y_sRM_train)
y_pred_test_NeuNet_sRM = model_NeuNet_sRM.predict(X_sRM_test)

In [296]:
print("\nTest Set Regression Models Performance without RM (average number of rooms per dwelling) field:\n")

print("** MSE (Mean Squared Error) **:")
print("Multiple Lineal Regression: ",mean_squared_error(y_sRM_test, y_pred_test_linear_sRM).round(4))
print("Polynomial Regression: ",mean_squared_error(y_sRM_test, y_pred_test_Poly_sRM).round(4))
print("Desiccion Tree: ", mean_squared_error(y_sRM_test, y_pred_test_decission_sRM).round(4))
print("Random Forest: ", mean_squared_error(y_sRM_test, y_pred_test_RanForest_sRM).round(4))
print("Neural Network: ", mean_squared_error(y_sRM_test, y_pred_test_NeuNet_sRM).round(4))

print("\n** R2 (Coefficient of determination) **:")
print("Multiple Lineal Regression: ",model_LinRegre_sRM.score(X_sRM_test, y_sRM_test).round(4))
print("Polynomial Regression: ",model_sRM_Poly.score(X_sRM_poly_test, y_sRM_test).round(4))
print("Desiccion Tree: ",model_DecTree_sRM.score(X_sRM_test, y_sRM_test).round(4))
print("Random Forest: ",model_RanForest_sRM.score(X_sRM_test, y_sRM_test).round(4))
print("Neural Network: ",model_NeuNet_sRM.score(X_sRM_test, y_sRM_test).round(4))


Test Set Regression Models Performance without RM (average number of rooms per dwelling) field:

** MSE (Mean Squared Error) **:
Multiple Lineal Regression:  30.8541
Polynomial Regression:  129.9387
Desiccion Tree:  29.6169
Random Forest:  22.4891
Neural Network:  27.3418

** R2 (Coefficient of determination) **:
Multiple Lineal Regression:  0.6294
Polynomial Regression:  -0.5605
Desiccion Tree:  0.6443
Random Forest:  0.7299
Neural Network:  0.6716


**Conclusió:**  
Tots els resultats empitjoren en comparació als obtinguts entre l'Execici1 i el 2, especielment la Regressió Polinòmica que deixaria de ser representativa per tenir un valor d'R2 negatiu. El motiu és que al treure del model el camp **RM**, que juntament amb el **LSTAT**, són els que ténen una correl.lació més forta amb el target **MEDV**, tots els resultats obtinguts amb els diferents models de regressió empitjoren.